system/lib/emmalloc.c

```diff
@@ -339,29 +347,29 @@ static void link_to_free_list(Region *freeRegion)
 static void dump_memory_regions()
 {
   ASSERT_MALLOC_IS_ACQUIRED();
-  Region *root = listOfAllRegions;
+  RootRegion *root = listOfAllRegions;
   MAIN_THREAD_ASYNC_EM_ASM(console.log('All memory regions:'));
   while(root)
   {
-    Region *r = root;
-    assert(debug_region_is_consistent(r));
-    uint8_t *lastRegionEnd = (uint8_t*)(((uint32_t*)root)[2]);
+    RootRegion *r = root;
+    assert(debug_region_is_consistent((Region*)r));
+    uint8_t *lastRegionEnd = root->endPtr;
     MAIN_THREAD_ASYNC_EM_ASM(console.log('Region block 0x'+($0>>>0).toString(16)+' - 0x'+($1>>>0).toString(16)+ ' ('+($2>>>0)+' bytes):'),
       r, lastRegionEnd, lastRegionEnd-(uint8_t*)r);
     while((uint8_t*)r < lastRegionEnd)
     {
       MAIN_THREAD_ASYNC_EM_ASM(console.log('Region 0x'+($0>>>0).toString(16)+', size: '+($1>>>0)+' ('+($2?"used":"--FREE--")+')'),
-        r, r->size, region_ceiling_size(r) == r->size);
+        r, r->size, region_ceiling_size((Region*)r) == r->size);
 
-      assert(debug_region_is_consistent(r));
-      uint32_t sizeFromCeiling = size_of_region_from_ceiling(r);
+      assert(debug_region_is_consistent((Region*)r));
+      uint32_t sizeFromCeiling = size_of_region_from_ceiling((Region*)r);
       if (sizeFromCeiling != r->size)
         MAIN_THREAD_ASYNC_EM_ASM(console.log('Corrupt region! Size marker at the end of the region does not match: '+($0>>>0)), sizeFromCeiling);
       if (r->size == 0)
         break;
-      r = next_region(r);
+      r = (RootRegion*)next_region((Region*)r);
```

Comment:
When a region block is being walked, the subsequent regions are not root regions. That is, line 354 and line 410 and line 1245 and line 1356 would read `Region *r = (Region*)root;`. That will get rid of the intermittent `(Region*)` casting in this function, and statically looks sound, since each `RootRegion` is also a `Region`.

i.e. `RootRegion` is an instance of a `Region`, but not all `Region`s are `RootRegion`s.

Reply:
Oh nice!  Much nicer.